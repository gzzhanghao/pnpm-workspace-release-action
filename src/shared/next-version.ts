import * as semver from 'semver';

export function getNextVersion(
  version: string,
  level: 'major' | 'minor' | 'patch',
  preVersion?: string,
  preid?: string,
): { version: string; preVersion?: string } {
  if (!preVersion) {
    if (!preid) {
      return {
        version: semver.inc(version, level)!,
      };
    }
    return {
      version: semver.inc(version, `pre${level}`, preid)!,
      preVersion: version,
    };
  }
  let bumpFromCurrent: string;
  let bumpFromPre: string;
  if (preid) {
    bumpFromCurrent = semver.inc(version, 'prerelease', preid)!;
    bumpFromPre = semver.inc(preVersion, `pre${level}`, preid)!;
  } else {
    bumpFromCurrent = version.split('-')[0];
    bumpFromPre = semver.inc(preVersion, level)!;
  }
  return {
    version: semver.gt(bumpFromCurrent, bumpFromPre)
      ? bumpFromCurrent
      : bumpFromPre,
    preVersion,
  };
}
