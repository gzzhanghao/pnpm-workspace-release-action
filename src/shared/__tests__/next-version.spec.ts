import { getNextVersion } from '../next-version';

const levels = ['major', 'minor', 'patch'] as const;

describe('calc', () => {
  test('1.0.0');
  test('1.0.0', undefined, 'alpha');

  test('1.0.1-alpha.0', '1.0.0', 'alpha');
  test('1.1.0-alpha.0', '1.0.0', 'alpha');
  test('2.0.0-alpha.0', '1.0.0', 'alpha');

  test('1.0.1-alpha.0', '1.0.0', 'beta');
  test('1.1.0-alpha.0', '1.0.0', 'beta');
  test('2.0.0-alpha.0', '1.0.0', 'beta');

  test('1.0.1-alpha.0', '1.0.0');
  test('1.1.0-alpha.0', '1.0.0');
  test('2.0.0-alpha.0', '1.0.0');
});

function test(version: string, preVersion?: string, preid?: string) {
  for (const level of levels) {
    it(`${version} + ${level} ${preVersion || '-'} ${preid || '-'}`, () => {
      expect(
        getNextVersion(version, level, preVersion, preid),
      ).toMatchSnapshot();
    });
  }
}
