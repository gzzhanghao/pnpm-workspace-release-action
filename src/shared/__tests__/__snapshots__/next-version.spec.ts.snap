// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calc 1.0.0 + major - - 1`] = `
{
  "version": "2.0.0",
}
`;

exports[`calc 1.0.0 + major - alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-alpha.0",
}
`;

exports[`calc 1.0.0 + minor - - 1`] = `
{
  "version": "1.1.0",
}
`;

exports[`calc 1.0.0 + minor - alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-alpha.0",
}
`;

exports[`calc 1.0.0 + patch - - 1`] = `
{
  "version": "1.0.1",
}
`;

exports[`calc 1.0.0 + patch - alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.0.1-alpha.0",
}
`;

exports[`calc 1.0.1-alpha.0 + major 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0",
}
`;

exports[`calc 1.0.1-alpha.0 + major 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-alpha.0",
}
`;

exports[`calc 1.0.1-alpha.0 + major 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-beta.0",
}
`;

exports[`calc 1.0.1-alpha.0 + minor 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0",
}
`;

exports[`calc 1.0.1-alpha.0 + minor 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-alpha.0",
}
`;

exports[`calc 1.0.1-alpha.0 + minor 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-beta.0",
}
`;

exports[`calc 1.0.1-alpha.0 + patch 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.0.1",
}
`;

exports[`calc 1.0.1-alpha.0 + patch 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.0.1-alpha.1",
}
`;

exports[`calc 1.0.1-alpha.0 + patch 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.0.1-beta.0",
}
`;

exports[`calc 1.1.0-alpha.0 + major 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0",
}
`;

exports[`calc 1.1.0-alpha.0 + major 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-alpha.0",
}
`;

exports[`calc 1.1.0-alpha.0 + major 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-beta.0",
}
`;

exports[`calc 1.1.0-alpha.0 + minor 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0",
}
`;

exports[`calc 1.1.0-alpha.0 + minor 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-alpha.1",
}
`;

exports[`calc 1.1.0-alpha.0 + minor 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-beta.0",
}
`;

exports[`calc 1.1.0-alpha.0 + patch 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0",
}
`;

exports[`calc 1.1.0-alpha.0 + patch 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-alpha.1",
}
`;

exports[`calc 1.1.0-alpha.0 + patch 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "1.1.0-beta.0",
}
`;

exports[`calc 2.0.0-alpha.0 + major 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0",
}
`;

exports[`calc 2.0.0-alpha.0 + major 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-alpha.1",
}
`;

exports[`calc 2.0.0-alpha.0 + major 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-beta.0",
}
`;

exports[`calc 2.0.0-alpha.0 + minor 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0",
}
`;

exports[`calc 2.0.0-alpha.0 + minor 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-alpha.1",
}
`;

exports[`calc 2.0.0-alpha.0 + minor 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-beta.0",
}
`;

exports[`calc 2.0.0-alpha.0 + patch 1.0.0 - 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0",
}
`;

exports[`calc 2.0.0-alpha.0 + patch 1.0.0 alpha 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-alpha.1",
}
`;

exports[`calc 2.0.0-alpha.0 + patch 1.0.0 beta 1`] = `
{
  "preVersion": "1.0.0",
  "version": "2.0.0-beta.0",
}
`;
